mangaki/mangaki/utils/recommendations.py

```diff
@@ -5,119 +5,142 @@
-    for work_id, choice in qs:
-        rated_works[work_id] = choice
-
-    willsee = set()
-    if user.profile.reco_willsee_ok:
-        banned_works = set()
-        for work_id in rated_works:
-            if rated_works[work_id] != 'willsee':
-                banned_works.add(work_id)
-            else:
-                willsee.add(work_id)
+        queryset = Rating.objects.all()
+
+    chrono.save('[%dQ] get interesting %d ratings' % (len(connection.queries), queryset.count()))
+
+    filename = 'ratings.pickle'
+    if algo_name == 'knn' or not os.path.isfile('ratings.pickle'):
+        ratings_pack = make_anonymous_data(queryset)
+        if algo_name != 'knn':
+            with open(filename, 'wb') as f:
+                pickle.dump(ratings_pack, f)
     else:
-        banned_works = set(rated_works.keys())
-
-    mangas = Work.objects.filter(category__slug='manga')
-
-    if editor == 'otototaifu':
-        mangas = mangas.filter(editor__title__in=['Ototo Manga', 'Taifu comics'])
-    elif editor != 'unspecified':
-        mangas = mangas.filter(editor__title__icontains=editor)
-    manga_ids = mangas.values_list('id', flat=True)
-
-    kept_works = None
-    if category == 'anime':
-        banned_works |= set(manga_ids)
-    elif category == 'manga':
-        kept_works = set(manga_ids)
-
-    chrono.save('[%dQ] retrieve her %d ratings' % (len(connection.queries), len(rated_works)))
-    final_works = Counter()
-    nb_ratings = {}
-    c = 0
-    neighbors = Counter()
-    for user_id, work_id, choice in Rating.objects.filter(work__in=rated_works.keys()).values_list('user_id', 'work_id', 'choice'):
-        c += 1
-        neighbors[user_id] += values[rated_works[work_id]] * rating_values[choice]
-
-    chrono.save('[%dQ] fill neighbors with %d ratings' % (len(connection.queries), c))
-
-    score_of_neighbor = {}
-    # print('Neighbors:')
-    # nbr = []
-    for user_id, score in neighbors.most_common(NB_NEIGHBORS):
-        # print(User.objects.get(id=user_id).username, score)
-        score_of_neighbor[user_id] = score
-        # nbr.append(user_id)
-    # print(nbr)
-
-    sum_ratings = Counter()
-    nb_ratings = Counter()
-    sum_scores = Counter()
-    i = 0
-    for work_id, user_id, choice in Rating.objects.filter(user__id__in=score_of_neighbor.keys()).exclude(choice__in=['willsee', 'wontsee']).values_list('work_id', 'user_id', 'choice'):
-        i += 1
-        if work_id in banned_works or (kept_works and work_id not in kept_works):
-            continue
-
-        sum_ratings[work_id] += rating_values[choice]
-        nb_ratings[work_id] += 1
-        sum_scores[work_id] += score_of_neighbor[user_id]
+        with open(filename, 'rb') as f:
+            ratings_pack = pickle.load(f)
+    X, y, nb_users, nb_works, encode_user, decode_user, encode_work, decode_work, interesting_works = ratings_pack
+
+    chrono.save('[%dQ] get all %d interesting ratings' % (len(connection.queries), queryset.count()))
+
+    algo = {
```

Comment:
This "dispatcher" dict should be defined somewhere else.
Moreover, I don't know if it's necessary a good idea to instantiate them.

Why wrap them into a function which can build them whenever we want? e.g.

```python
algos = {
   'knn': lambda: MangakiKNN(NB_NEIGHBORS, single_user=TRue, missing_is_mean=False)
}

algorithm = algos[algo_name]()
```